gcp/src/main/java/org/apache/iceberg/gcp/bigquery/BigQueryClientImpl.java

```diff
@@ -0,0 +1,481 @@
+            tablesStream
+                .parallel()
+                .filter(
+                    table -> {
+                      try {
+                        getTable(table.getTableReference());
+                      } catch (NoSuchTableException e) {
+                        return false;
+                      }
+                      return true;
+                    });
+      }
+
+      return tablesStream.collect(Collectors.toList());
+    } catch (IOException e) {
+      throw new RuntimeIOException(e);
+    }
+  }
+
+  @SuppressWarnings("FormatStringAnnotation")
+  private Dataset updateDataset(Dataset dataset) {
+    try {
+      HttpResponse response =
+          client
+              .datasets()
+              .update(
+                  Preconditions.checkNotNull(dataset.getDatasetReference()).getProjectId(),
+                  Preconditions.checkNotNull(dataset.getDatasetReference().getDatasetId()),
+                  dataset)
+              .setRequestHeaders(new HttpHeaders().setIfMatch(dataset.getEtag()))
+              .executeUnparsed();
+      if (response.getStatusCode() == HttpStatusCodes.STATUS_CODE_NOT_FOUND) {
+        throw new NoSuchNamespaceException(response.getStatusMessage());
+      }
+      return convertExceptionIfUnsuccessful(response).parseAs(Dataset.class);
+    } catch (IOException e) {
+      throw new RuntimeIOException(e);
+    }
+  }
+
+  /**
+   * Returns true when it is a BigQuery Metastore Iceberg table, defined by having the
+   * ExternalCatalogTableOptions object and a parameter of METADATA_LOCATION_PROP as part of its
+   * parameters map.
+   *
+   * @param table to check
+   */
+  private boolean isValidIcebergTable(Table table) {
+    return table.getExternalCatalogTableOptions() != null
+        && !table.getExternalCatalogTableOptions().isEmpty()
+        && table.getExternalCatalogTableOptions().getParameters() != null
+        && table
+            .getExternalCatalogTableOptions()
+            .getParameters()
+            .containsKey(BaseMetastoreTableOperations.METADATA_LOCATION_PROP)
+        && BaseMetastoreTableOperations.ICEBERG_TABLE_TYPE_VALUE.equalsIgnoreCase(
+            table
+                .getExternalCatalogTableOptions()
+                .getParameters()
+                .get(BaseMetastoreTableOperations.TABLE_TYPE_PROP));
+  }
+
+  private Table validateTable(Table table) {
+    if (!isValidIcebergTable(table)) {
+      throw new NoSuchIcebergTableException("This table is not a valid Iceberg table: %s", table);
+    }
+    return table;
+  }
+
+  /**
+   * Converts BigQuery generic API errors to Iceberg exceptions, *without* handling the
+   * resource-specific exceptions like NoSuchTableException, NoSuchNamespaceException, etc.
+   */
+  @SuppressWarnings("FormatStringAnnotation")
+  private HttpResponse convertExceptionIfUnsuccessful(HttpResponse response) throws IOException {
+    if (response.isSuccessStatusCode()) {
+      return response;
+    }
```

Comment:
Style: Iceberg adds newlines between control flow blocks and the following statments.

Reply:
Done everywhere.